ethereum/contracts/src/Announcements.sol

```diff
@@ -65,89 +79,172 @@ abstract contract HoprAnnouncementsEvents {
+     * @param from The address sending the tokens. Node or Safe contract could be the sender.
+     * If the sender is the Safe contract, it must be the one associated with the node.
+     * @param to The address receiving the tokens. Must be this contract's address.
+     * @param amount The amount of tokens sent. Must be equal to the key binding fee.
+     * @param userData The data sent along with the tokens. It must contain the key binding parameters.
+     *
      */
-    function bindKeysMaybeAnnounceSafe(
-        address selfAddress,
-        bytes32 ed25519_sig_0,
-        bytes32 ed25519_sig_1,
-        bytes32 ed25519_pub_key,
-        string calldata baseMultiaddr
+    function tokensReceived(
+        address, // operator not needed
+        address from,
+        address to,
+        uint256 amount,
+        bytes calldata userData,
+        bytes calldata // operatorData not needed
     )
         external
-        HoprMultiSig.onlySafe(selfAddress)
     {
-        _bindKeysInternal(selfAddress, ed25519_sig_0, ed25519_sig_1, ed25519_pub_key);
+        // don't accept any other tokens ;-)
+        if (msg.sender != address(TOKEN)) {
+            revert WrongToken();
+        }
+        // only accept tokens sent to this contract
+        if (to != address(this)) {
+            revert InvalidTokenRecipient();
+        }
+        // userData should contain at least 32 * 4 + 32 * 2 (for baseMultiaddr string) = 192 bytes:
+        if (userData.length < ERC777_HOOK_BIND_KEY_MAYBE_ANNOUNCE_SIZE) {
+            revert InvalidTokensReceivedUsage();
+        }
+        // decode userData to extract key binding parameters
+        (
+            address nodeAddress,
+            bytes32 ed25519_sig_0,
+            bytes32 ed25519_sig_1,
+            bytes32 ed25519_pub_key,
+            string memory baseMultiaddr
+        ) = abi.decode(userData, (address, bytes32, bytes32, bytes32, string));
+
+        // check the `from` address is the node or the Safe contract associated with the node
+        // accepted scenarios:
+        // 1) node without Safe pays for key binding: from == nodeAddress and expectedSafe == address(0)
+        // 2) node with Safe pays for key binding: from == expectedSafe && expectedSafe != address(0)
+        address expectedSafe = registry.nodeToSafe(nodeAddress);
+        if ((expectedSafe == address(0) && from != nodeAddress) || (expectedSafe != address(0) && from != expectedSafe))
+        {
+            revert InvalidTokenSender();
+        }
+
+        // perform optional announcement
         if (bytes(baseMultiaddr).length != 0) {
-            _announceInternal(selfAddress, baseMultiaddr);
+            _announceInternal(nodeAddress, baseMultiaddr);
         }
+        // perform idempotent key binding
+        (bool isNewInsertion,) = _bindKeysInternal(nodeAddress, ed25519_sig_0, ed25519_sig_1, ed25519_pub_key);
```

Comment:
binding must happen before announcement, otherwise the indexer won't be able to assign the announcement